dash/dash.py

```diff
@@ -2076,3 +2170,137 @@ def verify_url_part(served_part, url_part, part_name):
+                _, _, page_filename = page_filename.partition(pages + "/")
+                page_filename = page_filename.replace(".py", "").replace("/", ".")
+                module_name = ".".join([pages, page_filename])
+                page_module = importlib.import_module(module_name)
+
+                self.page_registry = _pages.PAGE_REGISTRY.copy()
+                if not self.page_registry[module_name]["supplied_layout"]:
+                    _validate.validate_pages_layout(module_name, page_module)
+                    self.page_registry[module_name]["layout"] = getattr(
+                        page_module, "layout"
+                    )
+
+    def _path_to_page(self, path_id):
+        path_variables = None
+        for page in self.page_registry.values():
+            if page["path_template"]:
+                template_id = page["path_template"].strip("/")
+                path_variables = _parse_path_variables(path_id, template_id)
+                if path_variables:
+                    return page, path_variables
+            if path_id == page["path"].strip("/"):
+                return page, path_variables
+        return {}, None
+
+    def enable_pages(self):
+        if not self.use_pages:
+            return
+        if self.pages_folder and not os.path.exists(self.config.pages_folder):
+            warnings.warn(
+                f"A folder called {self.pages_folder} does not exist.", stacklevel=2
+            )
+        if os.path.exists(self.config.pages_folder):
+            self._import_layouts_from_pages()
+
+        @self.server.before_first_request
+        def router():
+            @self.callback(
+                Output(_ID_CONTENT, "children"),
+                Output(_ID_STORE, "data"),
+                Input(_ID_LOCATION, "pathname"),
+                Input(_ID_LOCATION, "search"),
+                prevent_initial_call=True,
+            )
+            def update(pathname, search):
+                """
+                Updates dash.page_container layout on page navigation.
+                Updates the stored page title which will trigger the clientside callback to update the app title
+                """
+
+                # update page registry for pages that might have been added or changed in a callback
+                self.page_registry = _pages.PAGE_REGISTRY.copy()
+
+                query_parameters = _parse_query_string(search)
+                page, path_variables = self._path_to_page(
+                    self.strip_relative_path(pathname)
+                )
+
+                # get layout
+                if page == {}:
+                    if "pages.not_found_404" in self.page_registry:
+                        layout = self.page_registry["pages.not_found_404"]["layout"]
+                        title = self.page_registry["pages.not_found_404"]["title"]
+                    else:
+                        layout = html.H1("404")
```

Comment:
```suggestion
                        layout = html.H1("404 - Page not found")
```
A little friendlier to non-techy viewers :)